src/main/kotlin/net/ccbluex/liquidbounce/utils/block/placer/SupportFeature.kt

```diff
@@ -0,0 +1,140 @@
+import net.ccbluex.liquidbounce.utils.client.Chronometer
+import net.ccbluex.liquidbounce.utils.collection.Filter
+import net.ccbluex.liquidbounce.utils.math.sq
+import net.minecraft.block.Blocks
+import net.minecraft.util.math.BlockPos
+import net.minecraft.util.math.Direction
+import java.util.*
+import kotlin.Comparator
+
+// TODO support no wall range, proper reach calculations
+// wall range support could be done by taking the angle and, if the face is pointing to the player, exclude
+// TODO cache blocked / allowed spots
+/**
+ * Places blocks to support placements.
+ */
+class SupportFeature(val placer: BlockPlacer) : ToggleableConfigurable(placer, "Support", true) {
+
+    private val depth by int("Depth", 4, 1..12)
+    val delay by int("Delay", 500, 0..1000)
+
+    // what block helping blocks can be, by default just "trash blocks", meaning very common blocks
+    val filter by enumChoice("Filter", Filter.WHITELIST)
+    val blocks by blocks("Blocks", hashSetOf(Blocks.COBBLESTONE, Blocks.DIRT, Blocks.NETHERRACK))
+
+    // we don't have to consistently search, every once in a while is okay, see `delay`
+    val chronometer = Chronometer()
+
+    // positions we may not place at
+    val blockedPositions = hashSetOf<BlockPos>()
+
+    /**
+     * Currently finds the best path of blocks to support the placement of [targetPos] using Dijkstra's algorithm,
+     * the speed can possibly be improved by adding heuristics and making it an A* algorithm.
+     */
+    fun findSupport(targetPos: BlockPos): Set<BlockPos>? {
+        val rangeSq = placer.range.sq()
+
+        val openList = PriorityQueue<Node>(Comparator.comparingDouble { it.totalCost })
+        val closedList = hashSetOf<BlockPos>()
+
+        val startNode = Node(targetPos, null, 0.0)
+        openList.add(startNode)
+
+        while (!openList.isEmpty()) {
+            val currentNode = openList.poll()
+            closedList.add(currentNode.position)
+
+            // found a possible path
+            if (canPlace(currentNode.position)) {
+                return reconstructPath(currentNode)
+            }
+
+            for (direction in Direction.entries) {
+                val neighbor = currentNode.position.offset(direction)
+
+                // skip visited nodes
+                if (closedList.contains(neighbor)) {
+                    continue
+                }
+
+                if (
+                    // don't place helping blocks where the structure will be
+                    blockedPositions.contains(neighbor) ||
+
+                    // exclude blocks where the structure is...
+                    // this useless because we already search the shortest path under all structure blocks?
+                    placer.blocks.contains(neighbor) ||
+                    neighbor.manhattanDistanceTo(targetPos) > depth ||
+                    player.eyePos.squaredDistanceTo(neighbor.toCenterPos()) > rangeSq ||
+                    neighbor.isBlockedByEntities()
```

Comment:
Detekt also complained about this. You could extract those to variables: `isPlacementPositionIllegal`, `isTooDeep` and `canPlace`.

Reply:
It can in bad scenarios run over this function many times and while it would be cleaner and easier to read i think precomputing the conditions in variables would hurt performance, that's why i didn't to it here.